samza-core/src/main/java/org/apache/samza/operators/impl/WindowOperatorImpl.java

```diff
@@ -18,26 +18,379 @@
+
+    return newState;
+  }
+
+  private TriggerImplState getOrCreateTriggerWrapper(TriggerKey<K> triggerKey, Trigger<M> trigger) {
+    TriggerImplState wrapper = triggers.get(triggerKey);
+    if (wrapper != null) {
+      return wrapper;
+    }
+
+    TriggerImpl<M> triggerImpl = TriggerImpls.createTriggerImpl(trigger, clock);
+    TriggerContextImpl triggerContext = new TriggerContextImpl(triggerKey);
+    wrapper = new TriggerImplState(triggerKey, triggerImpl, triggerContext);
+    triggers.put(triggerKey, wrapper);
+
+    return wrapper;
+  }
+
+  /**
+   * Handles trigger firings, and propagates results to downstream operators.
+   */
+  private void onTriggerFired(TriggerKey<K> triggerKey, MessageCollector collector, TaskCoordinator coordinator) {
+    TriggerImplState wrapper = triggers.get(triggerKey);
+    WindowKey<K> windowKey = triggerKey.key;
+    WindowState<WV> state = store.get(windowKey);
+
+    if (state == null) {
+      return;
+    }
+
+    WindowPane<K, WV> paneOutput = computePaneOutput(triggerKey, state);
+    super.propagateResult(paneOutput, collector, coordinator);
+
+    // Handle accumulation modes.
+    if (window.getAccumulationMode() == AccumulationMode.DISCARDING) {
+      store.put(windowKey, null);
+    }
+
+    // Cancel all early triggers too when the default trigger fires. Also, clean all state for the key.
+    // note: We don't handle late arrivals yet, So, all arrivals are either early or on-time.
+    if (triggerKey.getType() == TriggerType.DEFAULT) {
+      cancelTrigger(triggerKey, true);
+      cancelTrigger(new TriggerKey(TriggerType.EARLY, triggerKey.getKey()), true);
+      WindowKey<K> key = triggerKey.key;
+      store.delete(key);
+    }
+
+    // Cancel non-repeating early triggers.
+    if (triggerKey.getType() == TriggerType.EARLY && !wrapper.isRepeating()) {
+      cancelTrigger(triggerKey, false);
+    }
+  }
+
+  /**
+   * Computes the pane output corresponding to a {@link TriggerKey} that fired.
+   */
+  private WindowPane<K, WV> computePaneOutput(TriggerKey<K> triggerKey, WindowState<WV> state) {
+    WindowKey<K> windowKey = triggerKey.key;
+    WV windowVal = state.getWindowValue();
+
+    //For session windows, we will create a new window key by using the time of the first message in the window as
+    //the paneId.
+    if (window.getWindowType() == WindowType.SESSION) {
+      windowKey = new WindowKey<>(windowKey.getKey(), Long.toString(state.getEarliestTime()));
+    }
+
+    //Make a defensive copy so that we are immune to further mutations on the collection
+    if (windowVal instanceof Collection) {
+      windowVal = (WV) new ArrayList<>((Collection<WV>) windowVal);
+    }
+
+    WindowPane<K, WV> paneOutput = new WindowPane<>(windowKey, windowVal, window.getAccumulationMode(), triggerKey.getType());
+    return paneOutput;
+  }
+
+  /**
+   * Cancels the firing of the {@link TriggerImpl} identified by this {@link TriggerKey} and optionally removes it.
```

Comment:
Thanks. Let's document what the implications of removing vs not removing a trigger are, and how the state for the window and trigger removal is related.